hal/inc/ble_hal.h

```diff
@@ -15,190 +15,747 @@
+ *                          service will be used instead. If NULL is provided on a central role and in response to a
+ *                          connection parameters update request, the peripheral request will be rejected.
+ *
+ * @returns     0 on success, system_error_t on error.
+ */
+int ble_gap_update_connection_params(uint16_t conn_handle, const hal_ble_conn_params_t* conn_params, void* reserved);
+
+/**
+ * Get connection parameters.
+ *
+ * @param[in]       conn_handle BLE connection handle.
+ * @param[in,out]   conn_params Pointer to a hal_ble_conn_params_t structure to be filled.
+ *
+ * @returns     0 on success, system_error_t on error.
+ */
+int ble_gap_get_connection_params(uint16_t conn_handle, hal_ble_conn_params_t* conn_params, void* reserved);
+
+/**
+ * Get the RSSI value of the specific BLE connection.
+ *
+ * @param[in]   conn_handle BLE connection handle.
+ *
+ * @returns     the RSSI value.
+ */
+int ble_gap_get_rssi(uint16_t conn_handle, void* reserved);
+
+/**
+ * Add a BLE 128-bits UUID service.
+ *
+ * @param[in]   type    BLE service type, either BLE_SERVICE_TYPE_PRIMARY or BLE_SERVICE_TYPE_SECONDARY.
+ * @param[in]   uuid    Pointer to BLE service UUID.
+ * @param[out]  handle  Service handle.
+ *
+ * @returns     0 on success, system_error_t on error.
+ */
+int ble_gatt_server_add_service(uint8_t type, const hal_ble_uuid_t* uuid, uint16_t* handle, void* reserved);
+
+/**
+ * Add a 128-bits UUID BLE Characteristic under a specific BLE Service.
+ *
+ * @note It is currently only possible to add a characteristic to the last added service (i.e. only sequential population is supported at this time).
+ *
+ * @param[in]   char_init   Pointer to the hal_ble_char_init_t structure.
+ * @param[out]  handles     Pointer to the hal_ble_char_handles_t structure.
+ *
+ * @returns     0 on success, system_error_t on error.
+ */
+int ble_gatt_server_add_characteristic(const hal_ble_char_init_t* char_init, hal_ble_char_handles_t* handles, void* reserved);
+
+/**
+ * Add descriptor under a specific BLE Characteristic.
+ *
+ * @note It is currently only possible to add a descriptor to the last added characteristic (i.e. only sequential population is supported at this time).
+ *
+ * @param[in]   desc_init   Pointer to the hal_ble_desc_init_t structure.
+ * @param[out]  handle      Handle of the descriptor.
+ *
+ * @returns     0 on success, system_error_t on error.
+ */
+int ble_gatt_server_add_descriptor(const hal_ble_desc_init_t* desc_init, uint16_t* handle, void* reserved);
+
+/**
+ * Set Characteristic value..
+ *
+ * @param[in]   value_handle    Characteristic value handle.
+ * @param[in]   buf             Pointer to the buffer that contains the data to be set.
+ * @param[in]   len             Length of the data to be set.
+ *
+ * @returns     Length of the data has been set.
+ */
+size_t ble_gatt_server_set_characteristic_value(uint16_t value_handle, const uint8_t* buf, size_t len, void* reserved);
```

Comment:
Return type should be signed: either `ssize_t` or `int` with negative values representing `system_error_t`.